src/apps/EDGE3D.hpp

```diff
@@ -0,0 +1,439 @@
+
+        const double tmpyzloc = tmpy*zloc;
+        const double tmpxzloc = tmpx*zloc;
+
+        const double yloctmpz = yloc*tmpz;
+        const double xloctmpz = xloc*tmpz;
+
+        const double jxx = Jxx(x, y, z, tmpyz, yloctmpz, tmpyzloc, yzloc);
+        const double jxy = Jxy(x, y, z, tmpyz, yloctmpz, tmpyzloc, yzloc);
+        const double jxz = Jxz(x, y, z, tmpyz, yloctmpz, tmpyzloc, yzloc);
+        const double jyx = Jyx(x, y, z, tmpxz, xloctmpz, tmpxzloc, xzloc);
+        const double jyy = Jyy(x, y, z, tmpxz, xloctmpz, tmpxzloc, xzloc);
+        const double jyz = Jyz(x, y, z, tmpxz, xloctmpz, tmpxzloc, xzloc);
+
+        double jinvxx, jinvxy, jinvxz,
+               jinvyx, jinvyy, jinvyz,
+               jinvzx, jinvzy, jinvzz,
+               detj_unfixed, detj, abs_detj, invdetj;
+
+        jacobian_inv(
+          jxx, jxy, jxz,
+          jyx, jyy, jyz,
+          jzx, jzy, jzz,
+          detj_cc, detj_tol,
+          jinvxx, jinvxy, jinvxz,
+          jinvyx, jinvyy, jinvyz,
+          jinvzx, jinvzy, jinvzz,
+          detj_unfixed, detj, abs_detj, invdetj);
+
+        const double detjwgts = wgts*abs_detj;
+
+        double ebasisx[EB] = {0};
+        edgebasis_x(ebasisx, tmpyz, yloctmpz, tmpyzloc, yzloc);
+
+        double ebasisy[EB] = {0};
+        edgebasis_y(ebasisy, tmpxz, xloctmpz, tmpxzloc, xzloc);
+
+        double dbasisz[EB] = {0};
+        curl_edgebasis_z(dbasisz, tmpz, zloc);
+
+        const double inv_abs_detj = 1./(abs_detj+ptiny);
+
+        double tebasisx[EB] = {0};
+        double tebasisy[EB] = {0};
+        double tebasisz[EB] = {0};
+
+        transform_edge_basis(
+          jinvxx, jinvxy, jinvxz,
+          jinvyx, jinvyy, jinvyz,
+          jinvzx, jinvzy, jinvzz,
+          ebasisx, ebasisy, ebasisz,
+          tebasisx, tebasisy, tebasisz);
+
+        double tdbasisx[EB] = {0};
+        double tdbasisy[EB] = {0};
+        double tdbasisz[EB] = {0};
+
+        transform_curl_edge_basis(
+          jxx, jxy, jxz,
+          jyx, jyy, jyz,
+          jzx, jzy, jzz,
+          inv_abs_detj,
+          dbasisx, dbasisy, dbasisz,
+          tdbasisx, tdbasisy, tdbasisz);
+
+        // the inner product: alpha*<w_i, w_j>
+        inner_product(
+          detjwgts*alpha,
+          tebasisx, tebasisy, tebasisz,
+          tebasisx, tebasisy, tebasisz,
+          matrix, true);
+
+         // the inner product: beta*<Curl(w_i), Curl(w_j)>
+        inner_product(
+          detjwgts*beta,
+          tdbasisx, tdbasisy, tdbasisz,
+          tdbasisx, tdbasisy, tdbasisz,
+          matrix, true);
+
+      }
+    }
+  }
+  sum[i] = 0.0;
+  for (int m = 0; m < EB; m++) {
+    Real_type check = 0.0;
+    for (int p = 0; p < EB; p++) {
+      check += edge_matrix[m*EB + p];
+    }
+    sum[i] += check;
+  }
+}
+*/
+///
+
+#ifndef RAJAPerf_Apps_EDGE3D_HPP
+#define RAJAPerf_Apps_EDGE3D_HPP
+
+#define NQ_1D 2
+
+#include "ahelper.hpp"
+
+__host__ __device__
+inline void edge_MpSmatrix(
```

Comment:
let's consider whether we think this should be in the macro or in a function

Reply:
Also, it should be changed because it doesn't compile.

Reply:
Ah those pesky non-gpu builds... ;)

fixed